### PR TITLE
[12.0] mig stock_picking_package_preparation_line

### PR DESCRIPTION
**Pull request by eLBati**
~~Depends on https://github.com/OCA/stock-logistics-workflow/pull/488~~

### CODE REVIEW

**Comment by eLBati** [comment]
Rebased after https://github.com/OCA/stock-logistics-workflow/pull/488

File: stock_picking_package_preparation_line/readme/DESCRIPTION.rst [hunk 1]
```diff
@@ -1,2 +1 @@
-This module adds a new class on **Package Preparation** to manage details
-related with stock move or not
+This module adds a new class on **Package Preparation** to manage related details with stock move or not
```
Comment (tafaRU) [inline stock_picking_package_preparation_line/readme/DESCRIPTION.rst:1]: why did you change it?

File: stock_picking_package_preparation_line/models/stock_picking_package_preparation.py [hunk 2]
```diff
@@ -171,8 +171,8 @@ def action_put_in_pack(self):
                 # preparation lines, updating lines would erase some fields
                 picking = picking.with_context(skip_update_line_ids=True)
                 picking.action_confirm()
-                # Show an error if a picking is not confirmed
-                if picking.state != 'confirmed':
+                # Show an error if a picking is not confirmed or assigned
+                if picking.state not in ('confirmed', 'assigned'):
```
Comment (rousseldenis) [inline stock_picking_package_preparation_line/models/stock_picking_package_preparation.py]: @eLBati As you changed this, you should include test.

File: stock_picking_package_preparation_line/readme/DESCRIPTION.rst (hunk 1 above)
Comment (eLBati) [reply stock_picking_package_preparation_line/readme/DESCRIPTION.rst:1]: @tafaRU to make it more clear

File: stock_picking_package_preparation_line/models/stock_picking_package_preparation.py (hunk 2 above)
Comment (eLBati) [reply stock_picking_package_preparation_line/models/stock_picking_package_preparation.py]: @rousseldenis what do you mean? That part is already covered

Comment (rousseldenis) [reply stock_picking_package_preparation_line/models/stock_picking_package_preparation.py]: Nope : https://codecov.io/gh/OCA/stock-logistics-workflow/compare/1e193d25dde1d9e23ed58bcd13335be02ab2a70c...79b5e82054339d234a06bf4613f573e08d731cc0/diff#D3-176

**Review by mrcast — COMMENTED** [review]
I did these steps:
Package Preparations - > Create
- I choose a partner
- in tab details I add product
Save and then Put in Pack
I obtain the following message:

"Cannot find a default location for picking type: Delivery Orders"

File: stock_picking_package_preparation_line/models/stock_picking_package_preparation.py (hunk 2 above)
Comment (eLBati) [reply stock_picking_package_preparation_line/models/stock_picking_package_preparation.py]: @rousseldenis added a test and discovered that change what not needed actually. Thanks

**Comment by eLBati** [comment]
@mrcast thanks. Fixed with a separate commit 6db69522d2ee0688a4f52e956055a57b8346f581 as it also happens in v11

**Review by mrcast — APPROVED** [review]
Functional test ok!
Thanks

**Comment by eLBati** [comment]
@tafaRU @rousseldenis what do you think?

File: stock_picking_package_preparation_line/tests/test_package_preparation_line.py [hunk 3]
```diff
@@ -82,12 +84,19 @@ def test_line_product_onchange(self):
...
+            # picking not confirmed nor assigned
+            self.preparation.action_put_in_pack()
+
     def test_move_and_picking_create_from_line(self):
         # Add a line in preparation
         # Test if a stock move and picking are created when "Put in Pack"
         # is used
         self._create_line(self.preparation, self.product2, 2.0)
         self.preparation.action_put_in_pack()
+        self.preparation.action_put_in_pack()
```
Comment (tafaRU) [inline stock_picking_package_preparation_line/tests/test_package_preparation_line.py]: Why did you add this line twice?

**Review by tafaRU — CHANGES_REQUESTED** [review]

File: stock_picking_package_preparation_line/readme/DESCRIPTION.rst (hunk 1 above)
Comment (tafaRU) [reply stock_picking_package_preparation_line/readme/DESCRIPTION.rst:1]: ok

File: stock_picking_package_preparation_line/tests/test_package_preparation_line.py (hunk 3 above)
Comment (eLBati) [reply stock_picking_package_preparation_line/tests/test_package_preparation_line.py]: @tafaRU removed, thanks

**Review by tafaRU — APPROVED** [review]

File: stock_picking_package_preparation_line/models/stock_picking_package_preparation.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+class StockPickingPackagePreparation(models.Model):
+
+    _inherit = 'stock.picking.package.preparation'
+
+    FIELDS_STATES = {'done': [('readonly', True)],
+                     'in_pack': [('readonly', True)],
+                     'cancel': [('readonly', True)]}
+
+    picking_type_id = fields.Many2one(
+        related='picking_ids.picking_type_id', readonly=True)
```
Comment (Nikul-Chaudhary) [inline stock_picking_package_preparation_line/models/stock_picking_package_preparation.py]: @eLBati no need to `readonly=True`

**Review by Nikul-Chaudhary — COMMENTED** [review]
Improve Code

File: stock_picking_package_preparation_line/models/stock_picking_package_preparation.py (hunk 4 above)
Comment (eLBati) [reply stock_picking_package_preparation_line/models/stock_picking_package_preparation.py]: @Nikul-Chaudhary removed

File: stock_picking_package_preparation_line/models/stock_picking_package_preparation.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+            if package_preparation_lines:
+                for line_vals in package_preparation_lines:
+                    line_vals['package_preparation_id'] = pack.id
+                    line_model.create(line_vals)
+
+    @api.model
+    @api.returns('self', lambda value: value.id)
+    def create(self, values):
+        values = self._update_line_vals(values)
+        pack = super(StockPickingPackagePreparation, self).create(values)
```
Comment (Nikul-Chaudhary) [inline stock_picking_package_preparation_line/models/stock_picking_package_preparation.py]: `return super(StockPickingPackagePreparation, self).create(values)`

**Review by Nikul-Chaudhary — COMMENTED** [review]
Improve Code

File: stock_picking_package_preparation_line/models/stock_picking_package_preparation.py (hunk 5 above)
Comment (eLBati) [reply stock_picking_package_preparation_line/models/stock_picking_package_preparation.py]: @Nikul-Chaudhary done

**Comment by eLBati** [comment]
Mergeable?

**Comment by rousseldenis** [comment]
Nearly. Can you just squash your commits?

**Comment by eLBati** [comment]
@rousseldenis  referring to 8beb57735c9d05b0b9c7a909021d3522cadb99a0 and 97acfd289ff1713e0fde57b88d43da43399a8330 I intentionally kept them separate, because 97acfd289ff1713e0fde57b88d43da43399a8330 is not strictly migration and could be ported to previous versions (v11 is affected)
